test_launch_ros/test/test_launch_ros/actions/test_ros_timer_action.py

```diff
@@ -0,0 +1,201 @@
+import sys
+import time
+import threading
+from functools import partial
+
+import launch
+import launch.event_handlers
+
+from launch_ros.actions import RosTimerAction
+
+import rclpy
+from rclpy.clock import Clock, ClockType
+
+from rosgraph_msgs.msg import Clock as ClockMsg
+from builtin_interfaces.msg import Time
+
+
+def test_multiple_launch_with_timers():
+    def generate_launch_description():
+        return launch.LaunchDescription([
+
+            launch.actions.ExecuteProcess(
+                cmd=[sys.executable, '-c', 'while True: pass'],
+            ),
+
+            RosTimerAction(
+                period=1.,
+                actions=[
+                    launch.actions.Shutdown(reason='Timer expired')
+                ]
+            ),
+        ])
+
+    ls = launch.LaunchService()
+    ls.include_launch_description(generate_launch_description())
+    assert 0 == ls.run()
+
+    ls = launch.LaunchService()
+    ls.include_launch_description(generate_launch_description())
+    assert 0 == ls.run()
+
+
+def _shutdown_listener_factory(reasons_arr):
+    return launch.actions.RegisterEventHandler(
+        launch.event_handlers.OnShutdown(
+            on_shutdown=lambda event, context: reasons_arr.append(event)
+        )
+    )
+
+
+def test_timer_action_sanity_check():
+    """Test that timer actions work (sanity check)."""
+    # This test is structured like test_shutdown_preempts_timers
+    # as a sanity check that the shutdown listener
+    # and other launch related infrastructure works as expected
+    shutdown_reasons = []
+
+    ld = launch.LaunchDescription([
+        launch.actions.ExecuteProcess(
+            cmd=[sys.executable, '-c', 'while True: pass'],
+        ),
+
+        RosTimerAction(
+            period=1.,
+            actions=[
+                launch.actions.Shutdown(reason='One second timeout')
+            ]
+        ),
+
+        _shutdown_listener_factory(shutdown_reasons),
+    ])
+
+    ls = launch.LaunchService()
+    ls.include_launch_description(ld)
+    assert 0 == ls.run()
+    assert shutdown_reasons[0].reason == 'One second timeout'
+
+
+def test_shutdown_preempts_timers():
+    shutdown_reasons = []
+
+    ld = launch.LaunchDescription([
+
+        launch.actions.ExecuteProcess(
+            cmd=[sys.executable, '-c', 'while True: pass'],
+        ),
+
+        RosTimerAction(
+            period=1.,
+            actions=[
+                launch.actions.Shutdown(reason='fast shutdown')
+            ]
+        ),
+
+        RosTimerAction(
+            period=2.,
+            actions=[
+                launch.actions.Shutdown(reason='slow shutdown')
+            ]
+        ),
+
+        _shutdown_listener_factory(shutdown_reasons),
+    ])
+
+    ls = launch.LaunchService()
+    ls.include_launch_description(ld)
+    assert 0 == ls.run()
+    assert len(shutdown_reasons) == 1
+    assert shutdown_reasons[0].reason == 'fast shutdown'
+
+
+def test_time_is_passing():
```

Comment:
nit: I think this test could be combined with earlier "sanity check". I'd keep the timer at 1 second to keep the duration of running tests shorter.

Reply:
I've combined them.